### PR TITLE
Tmz 1.6.0 maya calendar

### PR DESCRIPTION
**Pull request by t-m-z**
providing symbol tables for mayan calendars Long count, Tsolkin and Haab
providing a tool to convert between mayan calendar, julian date, julian calendar and gregorian calendar

### CODE REVIEW

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+  var oneCharacters = ['d', 'e', 'f', 'g'];
+  var fiveCharacters = ['a', 'b', 'c'];
+
+  var displays = <List<String>>[];
+
+  List<int> numbers = inputs.where((input) => input != null).map((input) {
+    var number = 0;
+    var display = <String>[];
+    input.toLowerCase().split('').forEach((segment) {
+      if (oneCharacters.contains(segment)) {
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:192]: `      if (oneCharacters.contains(segment)) {
Attach files by dragging & dropping, selecting or pasting them.
        number += 1;
        display.add(segment);
        return;
      }
      if (fiveCharacters.contains(segment)) {
        number += 5;
        display.add(segment);
        return;
      }
      return;

Please change to else if and then remove the useless return. `

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+  int dayCount = MayaLongCountToMayaDayCount(longCount);
+  if (dayCount == 0) return '4 Ahau';
+
+  dayCount = dayCount + 159;
+  dayCount = 1 + dayCount % 260;
+  print(dayCount % 20);
+  return (1 + (dayCount - 1) % 13).toString() + ' ' + maya_tzolkin[1 + (dayCount - 1) % 20];
+}
+
+String MayaDayCountToHaab(List longCount) {
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:254]: missing List data typ

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+    while (numberDec >= (mayaCalendarSystem[position - 1])) {
+      value++;
+      numberDec = numberDec - (mayaCalendarSystem[position - 1]);
+    }
+    result = result + _alphabet[value];
+  }
+  return result;
+}
+
+String MayaDayCountToTzolkin(List longCount) {
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:244]: missing List data typ

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+
+  dayCount = dayCount + 347;
+  dayCount = 1 + dayCount % 365;
+  return (1 + (dayCount - 1) % 20).toString() + ' ' + maya_haab[1 + (dayCount - 1) ~/ 20];
+}
+
+String MayaLongCount(List<int> longCount) {
+  if (MayaLongCountToMayaDayCount(longCount) == 0) return [0, 0, 0, 0, 13, 0, 0, 0, 0].join('.');
+
+  List<int> result = new List<int>();
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:266]: change to new List contructor <int>[]

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+  for (int i = longCount.length; i < 9; i++) result.add(0);
+  for (int i = 0; i < longCount.length; i++) result.add(longCount[i]);
+  if (result[4] == 0 && result[3] != 0) result[4] = 13;
+  return result.join('.');
+}
+
+int MayaLongCountToMayaDayCount(List<int> longCount) {
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:273]: private ?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+  int dayCount = 0;
+  longCount = longCount.reversed.toList();
+  for (int i = 0; i < longCount.length; i++) dayCount = dayCount + longCount[i] * mayaCalendarSystem[i];
+  return dayCount;
+}
+
+DateOutput MayaDayCountToJulianCalendar(int mayaDayCount){
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:280]: private ?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+  return JulianDateToJulianCalendar(MayaDayCountToJulianDate(mayaDayCount) * 1.0, true);
+}
+
+DateOutput MayaDayCountToGregorianCalendar(int mayaDayCount){
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:284]: private ?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+}
+
+int MayaDayCountToJulianDate(int mayaDayCount){
+  if (Prefs.getString('mayacalendar_correlation') == null || Prefs.getString('mayacalendar_correlation') == '')
+    return (mayaDayCount + CORRELATION_NUMBER[THOMPSON]);
+  else
+    return (mayaDayCount + CORRELATION_NUMBER[Prefs.getString('mayacalendar_correlation')]);
+}
+
+DateOutput JulianDateToGregorianCalendar(double jd, bool round){
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:295]: private ?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+    month = e - 13;
+    year = c - 4715;
+  }
+  if (round)
+    return DateOutput(day.toString().split('.')[0], MONTH[month], year.toString());
+  else  
+    return DateOutput(day.toString(), MONTH[month], year.toString());
+}
+
+DateOutput JulianDateToJulianCalendar(double jd, bool round){
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:320]: private ?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+    month = e - 13;
+    year = c - 4715;
+  }
+  if (round)
+    return DateOutput(day.toString().split('.')[0], MONTH[month], year.toString());
+  else
+    return DateOutput(day.toString(), MONTH[month], year.toString());
+}
+
+double JulianCalendarToJulianDate(String date){
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:344]: not used ?
private ?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+  if (month <= 2) {
+    year = year - 1;
+    month = month + 12;
+  }
+  int b = 0;
+
+  return (365.25 * (year + 4716)).floor() + (30.6001 * (month + 1)).floor() + day + b - 1524.5;
+}
+
+double GregorianCalendarToJulianDate(String date){
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:372]: not used ?
private ?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+  else
+    year = int.parse(dateList[2]);
+
+  if (month <= 2) {
+    year = year - 1;
+    month = month + 12;
+  }
+  int b = 0;
+
+  return (365.25 * (year + 4716)).floor() + (30.6001 * (month + 1)).floor() + day + b - 1524.5;
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:369]: 365.25,  4716, 30.6001
used several times. declare it as a constant?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+    }).toList(),
+    'numbers': longCountToList(input),
+    'vigesimal': vigesimal
+  };
+}
+
+List<int> longCountToList(int numberDec) {
+  if (numberDec == 0) return [0];
+
+  List<int> result = new List<int>();
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:160]: change to new List contructor <int>[]

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 14]
```diff
@@ -0,0 +1,398 @@
+// https://en.wikisource.org/wiki/An_Introduction_to_the_Study_of_the_Maya_Hieroglyphs/Chapter_3#fig19
+// https://en.wikipedia.org/wiki/Maya_calendar
+// https://en.wikipedia.org/wiki/Haab%CA%BC
+// https://en.wikipedia.org/wiki/Tzolk%CA%BCin
+// https://www.hermetic.ch/cal_stud/maya/chap2g.htm
+// https://rolfrost.de/maya.html
+
+import 'package:flutter/material.dart';
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:8]: remove unused imports

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+enum CORRELATION  {THOMPSON, SMILEY, WEITZEL}
+
+final THOMPSON_CORRELATION = 584283;
+final SMILEY_CORRELATION = 482699;
+final WEITZEL_CORRELATION = 774078;
+final THOMPSON = 'Thompson';
+final SMILEY = 'Smiley';
+final WEITZEL = 'Weitzel';
+
+final Map CORRELATION_NUMBER = {
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:21]: private ?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+  WEITZEL : WEITZEL_CORRELATION,
+};
+
+Map<String, String> CORRELATION_SYSTEMS = {
+  THOMPSON: 'Thompson (584283)',
+  SMILEY: 'Smiley (482699)',
+  WEITZEL: 'Weitzel (774078)',
+};
+
+final maya_longcount = {
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:33]: not used ?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+String MayaDayCountToHaab(List longCount) {
+  int dayCount = MayaLongCountToMayaDayCount(longCount);
+  if (dayCount == 0) return '8 Cumhu';
+
+  dayCount = dayCount + 347;
+  dayCount = 1 + dayCount % 365;
+  return (1 + (dayCount - 1) % 20).toString() + ' ' + maya_haab[1 + (dayCount - 1) ~/ 20];
+}
+
+String MayaLongCount(List<int> longCount) {
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:263]: private ?

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 18]
```diff
@@ -0,0 +1,398 @@
...
+  14: 'Ix',
+  15: 'Men',
+  16: 'Cib',
+  17: 'Caban',
+  18: 'Etznab',
+  19: 'Cauac',
+  20: 'Ahau',
+};
+
+final maya_haab = {
```
Comment (Mike-3) [inline lib/logic/tools/science_and_technology/maya_calendar.dart:68]: private ?

File: lib/widgets/searchstrings/searchstrings_de.dart [hunk 19]
```diff
@@ -402,6 +403,9 @@ const SEARCHSTRING_DE_SYMBOL_MANDALORIAN = 'mandalorianisch';
 const SEARCHSTRING_DE_SYMBOL_MARAIN = '';
 const SEARCHSTRING_DE_SYMBOL_MARAIN_V2 = '';
 const SEARCHSTRING_DE_SYMBOL_MATORAN = '';
+const SEARCHSTRING_DE_SYMBOL_MAYA_CALENDAR_HAAB = 'kalender';
+const SEARCHSTRING_DE_SYMBOL_MAYA_CALENDAR_LONGCOUNT = 'lange zählung';
```
Comment (Mike-3) [inline lib/widgets/searchstrings/searchstrings_de.dart:407]: change diacritic

File: lib/logic/tools/science_and_technology/maya_calendar.dart (hunk 2 above)
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/maya_calendar.dart:254]: is used in widget

File: lib/logic/tools/science_and_technology/maya_calendar.dart (hunk 5 above)
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/maya_calendar.dart:273]: used in widget

File: lib/logic/tools/science_and_technology/maya_calendar.dart (hunk 6 above)
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/maya_calendar.dart:280]: used in widget

File: lib/logic/tools/science_and_technology/maya_calendar.dart (hunk 7 above)
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/maya_calendar.dart:284]: used in widget

File: lib/logic/tools/science_and_technology/maya_calendar.dart (hunk 8 above)
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/maya_calendar.dart:295]: yes

Comment (t-m-z) [reply lib/logic/tools/science_and_technology/maya_calendar.dart:295]: but can in future be replaced with date_Utils from calendar conversion

File: lib/logic/tools/science_and_technology/maya_calendar.dart (hunk 9 above)
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/maya_calendar.dart:320]: yes - but ... in future will this be change due to calender_conversion

File: lib/logic/tools/science_and_technology/maya_calendar.dart (hunk 10 above)
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/maya_calendar.dart:344]: done

File: lib/widgets/tools/science_and_technology/maya_calendar.dart [hunk 20]
```diff
@@ -0,0 +1,256 @@
+import 'dart:math';
+
+import 'package:flutter/material.dart';
+import 'package:gc_wizard/i18n/app_localizations.dart';
+import 'package:gc_wizard/logic/tools/science_and_technology/maya_calendar.dart';
+import 'package:gc_wizard/theme/theme.dart';
+import 'package:gc_wizard/widgets/common/base/gcw_divider.dart';
```
[the remark this answers is not in the document]
Comment (t-m-z) [reply lib/widgets/tools/science_and_technology/maya_calendar.dart:7]: done

File: lib/logic/tools/science_and_technology/maya_calendar.dart [hunk 21]
```diff
@@ -0,0 +1,398 @@
...
+  3: 'tun',
+  4: 'katun',
+  5: 'baktun',
+  6: 'pictun',
+  7: 'kalabtun',
+  8: 'kinchiltun',
+  9: 'alatun'
+};
+
+final maya_tzolkin = {
```
[the remark this answers is not in the document]
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/maya_calendar.dart:45]: done

File: lib/logic/tools/science_and_technology/maya_calendar.dart (hunk 17 above)
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/maya_calendar.dart:263]: used in widget

File: lib/widgets/searchstrings/searchstrings_de.dart (hunk 19 above)
Comment (t-m-z) [reply lib/widgets/searchstrings/searchstrings_de.dart:407]: done

**Review by Mike-3 — APPROVED** [review]
review finished

**Comment by t-m-z** [comment]
Merge conflicts solved

File: assets/i18n/en.json [hunk 22]
```diff
@@ -2598,6 +2613,11 @@
   "symboltables_marain_v2_description": "Sci-Fi language created by Iain M. Banks for his books, Version 2",
   "symboltables_matoran_title": "Matoran",
   "symboltables_matoran_description": "From 'Lego Bionicle'",
+  "symboltables_maya_calendar_haab_codices_title": "Maya Calendar Haab, Codices",
+  "symboltables_maya_calendar_haab_inscripts_title": "Maya Calendar Haab, Inscription",
+  "symboltables_maya_calendar_longcount_title" : "Maya Kalender Long count",
```
Comment (S-Man42) [inline assets/i18n/en.json:2618]: Typo: Kalender

**Comment by S-Man42** [comment]
I am thinking of moving this section into maya numbers or date calculations? what to you think?

**Comment by t-m-z** [comment]
I would prefer Date calculations

Sent from MailDroid

-----Original Message-----

File: assets/i18n/en.json (hunk 22 above)
Comment (t-m-z) [reply assets/i18n/en.json:2618]: done

**Comment by t-m-z** [comment]
Moved to Date and Time selection
correct typo
PR done

**Pull request merged**